taichi/ir/frontend_ir.cpp

```diff
@@ -9,10 +9,14 @@
 
 namespace taichi::lang {
 
-#define TI_ASSERT_TYPE_CHECKED(x)                       \
-  TI_ASSERT_INFO(x->ret_type != PrimitiveType::unknown, \
-                 "[{}] was not type-checked",           \
-                 ExpressionHumanFriendlyPrinter::expr_to_string(x))
+#define TI_ASSERT_TYPE_CHECKED(x)                                        \
+  if (x->ret_type == PrimitiveType::unknown) {                           \
+    ErrorEmitter(                                                        \
+        TaichiTypeError(), x.expr.get(),                                 \
+        fmt::format("[{}] was not type-checked",                         \
+                    ExpressionHumanFriendlyPrinter::expr_to_string(x))); \
+  } else {                                                               \
```

Comment:
Is the `else` scope necessary here?

Reply:
I was hoping that the `if` statement in the macro function won't mess the context of the caller up, i.e. altering the control flow in an amazing way when used in certain strange corner of the code. It seems that a do-while(false) is the standard practice to achieve this~